katsdptelstate/aio/redis.py

```diff
@@ -178,103 +130,78 @@ def get(self, key: bytes) -> Union[
-                            value, timestamp = utils.split_timestamp(data)
-                            timeout = yield MutableKeyUpdate(key, value, timestamp)
-                        else:
-                            raise RuntimeError('Unhandled key type {}'.format(key_type))
-
-    def dump(self, key: bytes) -> Optional[bytes]:
-        return self.client.dump(key)
+    async def monitor_keys(self, keys: Iterable[bytes]) -> AsyncGenerator[KeyUpdateBase, None]:
+        # Refer to katsdptelstate.redis.RedisBackend for details of the protocol
+        while True:
+            receiver = aioredis.pubsub.Receiver()
+            await self.client.subscribe(*(receiver.channel(b'update/' + key) for key in keys))
+            try:
+                # Condition may have been satisfied while we were busy subscribing
+                logger.debug('Subscribed to channels, notifying caller to check')
+                yield KeyUpdateBase()
+                async for channel, data in receiver.iter():
+                    assert channel.name.startswith(b'update/')
+                    key = channel.name[7:]
+                    logger.debug('Received update on channel %r', key)
+                    # First check if there is a key type byte in the message.
+                    try:
+                        key_type = KeyType(data[0])
+                    except (ValueError, IndexError):
+                        # It's the older format lacking a key type byte
+                        key_type2 = await self.key_type(key)
+                        if key_type2 is None:
+                            continue     # Key was deleted before we could retrieve the type
+                        key_type = key_type2
+                    else:
+                        data = data[1:]  # Strip off the key type
+                    if key_type == KeyType.IMMUTABLE:
+                        yield ImmutableKeyUpdate(key, data)
+                    elif key_type == KeyType.INDEXED:
+                        # Encoding is <sub-key-length>\n<sub-key><value>
+                        newline = data.find(b'\n')
+                        if newline == -1:
+                            raise RuntimeError('Pubsub message missing newline')
+                        sub_key_len = int(data[:newline])
+                        sub_key = data[newline + 1 : newline + 1 + sub_key_len]
+                        value = data[newline + 1 + sub_key_len :]
+                        yield IndexedKeyUpdate(key, sub_key, value)
+                    elif key_type == KeyType.MUTABLE:
+                        value, timestamp = utils.split_timestamp(data)
+                        yield MutableKeyUpdate(key, value, timestamp)
+                    else:
+                        raise RuntimeError('Unhandled key type {}'.format(key_type))
+            finally:
+                await asyncio.shield(self.client.unsubscribe(*(b'update/' + key for key in keys)))
```

Comment:
Any reason to worry that shielding from cancellation might end up blocked ?

Reply:
I ended up changing how this works in later commits.